### ISSUE TITLE
Phase 4 종합 결과 + 부족한 부분 (paper-readiness gap analysis)

### ISSUE THREAD

**Issue by kimshhhhh** [kimshhhhh/consistency-reasoning#15]
## 0. TL;DR

10개 실험 (3 GPU 병렬, 약 30시간) 후 **유일한 winning recipe**: `SFT → GRPO sequential with boundary data, 3ep`. SFT alone(overall 0.5995)을 통계적으로(p=0.047) 이긴 유일한 config. 다만 boundary zone에선 borderline (p=0.053). **Pure GRPO from scratch는 어떤 hyperparameter 조합으로도 SFT 못 이김** — RLVR의 본질적 한계 ([C1 NeurIPS 2025 Oral](https://openreview.net/forum?id=4OsgYD7em5)) 확정.

지금 결과로 **EMNLP/ACL main conference 가긴 약함**. 다음 섹션의 \"부족한 부분\" 채워야 paper-ready.

---

## 1. 실험 매트릭스 (10 configs)

| Tag | Base | Data | Epochs | β | LR | G | 비고 |
|---|---|---|---|---|---|---|---|
| `boundary` (1ep) | base | boundary | 1 | 0.04 | 1e-6 | 2 | 원본 |
| `random` | base | random | 1 | 0.04 | 1e-6 | 2 | 데이터 ablation |
| `all` | base | all | 1 | 0.04 | 1e-6 | 2 | 데이터 ablation |
| `boundary_e3` | base | boundary | 3 | 0.04 | 1e-6 | 2 | epoch 매칭 |
| `boundary_lowbeta_e3` | base | boundary | 3 | **0.01** | 1e-6 | 2 | KL 줄임 |
| `boundary_highlr_e3` | base | boundary | 3 | 0.04 | **3e-6** | 2 | LR 3배 |
| `boundary_sftinit_e3` ★ | **SFT** | boundary | 3 | 0.04 | 1e-6 | 2 | **winning recipe** |
| `random_sftinit_e3` | SFT | random | 3 | 0.04 | 1e-6 | 2 | boundary ablation |
| `boundary_sftinit_lowbeta_e3` | SFT | boundary | 3 | 0.01 | 1e-6 | 2 | combo (n.s.) |
| ❌ `boundary_g4_e3` | base | boundary | 3 | 0.04 | 1e-6 | **4** | OOM (24GB 한계) |

---

## 2. 결과 (pass@1, 500 problems)

```
Method                                       Overall  Inner  Boundary  Outer
─────────────────────────────────────────────────────────────────────────────
Phase 2 SFT (target)                         0.5995  0.9613  0.4643   0.0163
─────────────────────────────────────────────────────────────────────────────
GRPO from scratch (base = Qwen2.5-7B):
  boundary (1ep, β=0.04)                     0.5962  0.9704  0.4467   0.0167
  random (1ep)                               0.5911  0.9688  0.4261   0.0193
  all (1ep)                                  0.5909  0.9675  0.4361   0.0168
  boundary_e3 (3ep)                          0.5918  0.9670  0.4440   0.0176
  boundary_lowbeta_e3 (β=0.01)               0.5952  0.9708  0.4525   0.0225
  boundary_highlr_e3 (lr=3e-6)               0.5906  0.9707  0.4568   0.0193
─────────────────────────────────────────────────────────────────────────────
SFT-init + GRPO (base = Phase 2 SFT):
  random_sftinit_e3                          0.6031  0.9605  0.4561   0.0137
  boundary_sftinit_e3                  ★     0.6052  0.9612  0.4658   0.0162  ← winner
  boundary_sftinit_lowbeta_e3                0.5991  0.9605  0.4677   0.0173
```

---

### ✓ 통계적으로 유의 (p<0.05)
| 비교 | 영역 | p | n |
|---|---|---|---|
| **`sftinit_e3` > SFT alone** | overall | **0.0470** | 500 |
| boundary > random (1ep) | boundary zone | 0.0417 | — |
| boundary > all (1ep) | boundary zone | 0.0238 | — |

### Borderline (0.05 < p < 0.10)
- `sftinit_e3` > SFT, boundary zone: p=**0.0533** (n=89, statistical power 부족 가능성)
- `sftinit_e3` > GRPO(boundary, 3ep), boundary: p=0.0940

### ✗ 유의 안 됨
- FINAL (lowbeta combo) > SFT: boundary p=0.1822, overall p=0.3551
- SFT+GRPO(random) > SFT: p=0.0977
- 다양한 다른 ablation들

---

## 4. Paper로 쓸 수 있는 5개 핵심 발견

| # | 발견 | 증거 강도 |
|---|---|---|
| 1 | GRPO from scratch는 SFT를 못 이긴다 | ★★★★★ (10 configs 모두 fail) |
| 2 | **SFT → GRPO sequential은 SFT alone을 이긴다** | ★★★★ (overall p=0.047) |
| 3 | Boundary > Random/All consistently | ★★★★ (p<0.05 in multiple comparisons) |
| 4 | **Random data + GRPO는 SFT를 망가뜨린다** (boundary zone) | ★★★ (numerical, n.s. statistically) |
| 5 | Low β는 outer zone exploration에 도움 | ★★★ (numerical) |

---

### 5.1 통계적 power 부족
- **Boundary zone n=89~96**: 작은 샘플로 인한 unstable p-value
- 단일 seed → run-to-run variance 측정 안 됨 (실제 효과 vs noise 구분 안 됨)
- **필요**: 최소 3 seeds × main configs (sftinit_e3, sftinit_random_e3, GRPO baseline) 재실행. 약 30~40시간.

### 5.2 외부 타당성 (External validity)
- **단일 base 모델**: Qwen2.5-7B-Instruct만 사용 → 결과가 모델-specific인지 불명
- **단일 데이터셋**: MATH만 사용 → 다른 reasoning 도메인 (코드, 논리) 일반화 안 됨
- **필요**: 최소 1개 추가 모델 (예: Llama-3-8B) + 1개 추가 dataset (예: GSM8K)에서 재현

### 5.3 평가 메트릭 빈약
지금 pass@1 단일 메트릭만 사용. 부족한 것:
- **CoT-Pass@K** ([C2 paper](https://arxiv.org/abs/2506.14245)) — 답+추론 둘 다 평가. RLVR 효과를 더 잘 포착
- **Calibration (ECE)** — DCPO ([M-B1](https://arxiv.org/html/2603.09117))는 \"RLVR이 over-confidence 유발\" 주장. 우리 모델 confidence 측정 안 함
- **Boundary→inner graduation rate** — 측정은 했지만 deep analysis 없음 (SFT 28% vs GRPO 7% 차이의 원인 분석 부재)
- **메타인지 평가 5축** ([M-A1 LRM metacognition paper](https://openreview.net/forum?id=JGG9EdHyZc)) — difficulty awareness, confidence adjustment 등

#### (a) Reward 단순함
- Binary 0/1 reward → C1의 \"sharpening only\" 한계의 자연스러운 귀결
- **개선 방향**: 
  - Confidence-aware reward (Brier/Log score; [M-B2 EMNLP 2025](https://aclanthology.org/2025.emnlp-main.1385.pdf))
  - Process-level reward ([D2 GRPO-VPS](https://arxiv.org/html/2604.20659))
  - Self-consistency reward (TTRL 방식; [B2](https://arxiv.org/pdf/2504.16084))

#### (b) Boundary 정의 단순함
- Threshold (`pass@32 ∈ [0.1, 0.8]`) 임의 선택. 다른 threshold에서 robust한지 unclear
- Online vs offline boundary 분류 비교 안 됨 ([A1 NeurIPS 2025 paper](https://arxiv.org/abs/2506.05316)는 online으로 함)
- **개선 방향**: threshold sensitivity analysis + adaptive online boundary 비교

#### (c) Sequential SFT-GRPO 학습이 진짜 \"GRPO 효과\"인지 명확하지 않음
- SFT(0.4643) → SFT+GRPO(0.4658) 차이가 0.0015 — 매우 작음
- 같은 cgbd SFT 데이터로 SFT를 1 epoch 더 (4 epoch 총) 학습하면 비슷한 향상 가능성?
- **필요**: \"SFT 4 epoch (no GRPO)\" baseline 추가해야 GRPO의 incremental 가치 입증

### 5.5 Negative result의 framing
- **GRPO < SFT**라는 결과가 paper의 약점이 됨. C1 같이 \"RLVR 한계\"를 main 메시지로 가는 길 vs \"sequential이 valid\"를 main 메시지로 가는 길 둘 다 가능하지만, 둘 다 incremental
- **필요**: 우리만의 unique angle 추가. 현재 후보:
  - **메타인지 framing** (Issue #14) — \"data selection IS metacognition\"
  - **Process reward + boundary** combo — Issue 시작점
  - **Sequential pipeline의 quantitative analysis** — 어떤 경우 SFT→GRPO가 효과 있는지 fine-grained 분석

### 5.6 정량적 한계 분석 부재
지금까지 한 분석은 결과 비교 표만. 빠진 것:
- **어떤 종류의 boundary 문제에서 GRPO가 효과 있는지**: 도형, 대수, 수론 등 카테고리별 분석
- **틀린 답들의 분포**: 모델이 \"가까운 답\" 내는지 vs \"전혀 다른 답\"인지
- **Solution length 변화**: GRPO 후 풀이가 짧아지나/길어지나 (response collapse 우려)
- **Token-level 변화**: SFT vs GRPO 모델이 토큰을 어떻게 다르게 사용하는지 (entropy, top-k diversity)

### 5.7 Reproducibility
- 4개 버그 fix (issue #12 1차 댓글 + 2차 댓글) 했지만 코드 cleanup 안 됨
- README, requirement.txt 정리 안 됨
- HP search log 체계적 보관 안 됨

---

### 🔴 P0 (paper 갈 거면 필수)
1. **Multi-seed 재실행** (3 seeds × main 3 configs) — 통계 신뢰도 확보 [~30h compute]
2. **SFT 4 epoch baseline** — GRPO incremental 가치 분리 [~3h]
3. **CoT-Pass@K 평가 추가** [~6h, 코드 작성 + eval]

### 🟡 P1 (강한 paper용)
4. **Calibration (ECE) 측정** — RLVR over-confidence 가설 검증 [~3h]
5. **Confidence-aware reward 실험** — 메타인지 component 추가 [~10h, 새 reward function 구현 + 1 train run]
6. **Threshold sensitivity** — boundary 정의 robust한지 [~6h, 다른 threshold로 1~2 train]

### 🟢 P2 (Future work)
7. 다른 모델 (Llama-3-8B 등) 재현
8. 다른 dataset (GSM8K, AIME 등) 재현
9. Process-level reward 실험
10. 정량적 case study (어떤 문제 종류에서 GRPO 효과 있는지)

---

### 현재 결과만으로는: **Workshop short paper 수준**
- Engineering contribution: 4 bugs in trl 0.14.0 + Qwen2 stack ✓
- Empirical: \"sequential SFT+GRPO works, pure GRPO doesn't\" — 알려진 직관 confirm
- Negative-result-heavy: 9개 GRPO config가 SFT 못 이김

### Multi-seed + CoT-Pass@K + calibration 추가 시: **EMNLP findings / NAACL findings 수준**
- Statistical rigor 확보
- 새 메트릭으로 hidden trend 발견 가능성

### 메타인지 reward 추가 시: **Main conference 가능성**
- 새 method (boundary + metacognition reward) → novelty
- C1 한계 \"극복\" framing 가능
- 약 1~2개월 추가 노력 필요

---

**Related**: #12 (initial Phase 4 results + 4 bug fixes), #13 (RLVR/curriculum related work, 12 papers), #14 (metacognition reframing, 9 papers)
